Benchmark Reporting - Deadlock/Getting Stuck Reading From Process.StandardOutput (NetCore 2.1)

#### Update (7/15/2018 @ 7:25 PM EST)
Decoupled my Benchmark project from a local project and instead added reference to the NuGet it corresponds to.

In addition to that, I installed 2.302 SDK (`NetCore 2.1.2`).

Standard Sleep benchmark is working again.

Moving to BenchmarkDotNet nightlies (BDN v0.10.14.685, Core v0.10.14.508).

#### Update (7/15/2018 @ 8:07 PM EST)
Nothing seems to work.

Thread stuck here:
![image](https://user-images.githubusercontent.com/1414138/42739871-1b3c0d3e-886d-11e8-8251-f98258663e0b.png)

Found the exception in background:
![image](https://user-images.githubusercontent.com/1414138/42739968-4fa063bc-886e-11e8-9c86-8285b17b5b94.png)

Not sure how I can help but the file it's stuck on seems to be the .log.

I allowed continuation found the code frozen in one spot - continuously frozen in one spot:
![image](https://user-images.githubusercontent.com/1414138/42740075-e8a4afe0-886f-11e8-9178-7e21727d28f9.png)

```
Not Flagged	>	40620	1	Main Thread	Main Thread  

System.Private.CoreLib.dll!System.IO.FileStream.ReadFileNative    

[Managed to Native Transition]   

System.Private.CoreLib.dll!System.IO.FileStream.ReadFileNative(Microsoft.Win32.SafeHandles.SafeFileHandle handle, System.Span<byte> bytes, System.Threading.NativeOverlapped* overlapped, out int errorCode) Line 1190  
System.Private.CoreLib.dll!System.IO.FileStream.ReadNative(System.Span<byte> buffer) Line 507   
System.Private.CoreLib.dll!System.IO.FileStream.ReadSpan(System.Span<byte> destination) Line 433   
System.Private.CoreLib.dll!System.IO.FileStream.Read(byte[] array, int offset, int count) Line 302   
System.Private.CoreLib.dll!System.IO.StreamReader.ReadBuffer() Line 637  
System.Private.CoreLib.dll!System.IO.StreamReader.ReadLine() Line 812  
BenchmarkDotNet.dll!BenchmarkDotNet.Loggers.SynchronousProcessOutputLoggerWithDiagnoser.ProcessInput()  
BenchmarkDotNet.dll!BenchmarkDotNet.Toolchains.DotNetCli.DotNetCliExecutor.Execute(BenchmarkDotNet.Running.BenchmarkCase benchmarkCase, BenchmarkDotNet.Running.BenchmarkId benchmarkId, BenchmarkDotNet.Loggers.ILogger logger, BenchmarkDotNet.Toolchains.ArtifactsPaths artifactsPaths, BenchmarkDotNet.Diagnosers.IDiagnoser diagnoser, string executableName, BenchmarkDotNet.Configs.IConfig config, BenchmarkDotNet.Characteristics.IResolver resolver)   
BenchmarkDotNet.dll!BenchmarkDotNet.Toolchains.DotNetCli.DotNetCliExecutor.Execute(BenchmarkDotNet.Toolchains.Parameters.ExecuteParameters executeParameters)  
BenchmarkDotNet.dll!BenchmarkDotNet.Running.BenchmarkRunner.Execute(BenchmarkDotNet.Loggers.ILogger logger, BenchmarkDotNet.Running.BenchmarkCase benchmarkCase, BenchmarkDotNet.Running.BenchmarkId benchmarkId, BenchmarkDotNet.Toolchains.IToolchain toolchain, BenchmarkDotNet.Toolchains.Results.BuildResult buildResult, BenchmarkDotNet.Configs.IConfig config, BenchmarkDotNet.Characteristics.IResolver resolver)  
BenchmarkDotNet.dll!BenchmarkDotNet.Running.BenchmarkRunner.RunCore(BenchmarkDotNet.Running.BenchmarkCase benchmarkCase, BenchmarkDotNet.Running.BenchmarkId benchmarkId, BenchmarkDotNet.Loggers.ILogger logger, BenchmarkDotNet.Configs.ReadOnlyConfig config, BenchmarkDotNet.Characteristics.IResolver resolver, BenchmarkDotNet.Toolchains.Results.BuildResult buildResult)  
BenchmarkDotNet.dll!BenchmarkDotNet.Running.BenchmarkRunner.Run(BenchmarkDotNet.Running.BenchmarkRunInfo benchmarkRunInfo, System.Collections.Generic.Dictionary<BenchmarkDotNet.Running.BenchmarkCase, (BenchmarkDotNet.Running.BenchmarkId benchmarkId, BenchmarkDotNet.Toolchains.Results.BuildResult buildResult)> buildResults, BenchmarkDotNet.Characteristics.IResolver resolver, BenchmarkDotNet.Loggers.ILogger logger, System.Collections.Generic.List<string> artifactsToCleanup, string rootArtifactsFolderPath, ref BenchmarkDotNet.Horology.StartedClock runChronometer)
BenchmarkDotNet.dll!BenchmarkDotNet.Running.BenchmarkRunner.Run(BenchmarkDotNet.Running.BenchmarkRunInfo[] benchmarkRunInfos, BenchmarkDotNet.Configs.IConfig commonSettingsConfig)  
BenchmarkDotNet.dll!BenchmarkDotNet.Running.BenchmarkRunner.Run(BenchmarkDotNet.Running.BenchmarkRunInfo benchmarkRunInfo)  
CookedRabbit.Core.Benchmark.dll!CookedRabbit.Core.Benchmark.Program.Main(string[] args)  
CookedRabbit.Core.Benchmark.dll!CookedRabbit.Core.Benchmark.Program.<Main>(string[] args)  
```
Guessing this needs a timeout?
```
BenchmarkDotNet.dll!BenchmarkDotNet 
.Loggers
.SynchronousProcessOutputLoggerWithDiagnoser
.ProcessInput()
```

`BenchmarkDotNet.Loggers.SynchronousProcessOutputLoggerWithDiagnoser Line 38`

```csharp
internal void ProcessInput()
{
    string line;
    while ((line = process.StandardOutput.ReadLine()) != null)
    {
        logger.WriteLine(LogKind.Default, line);

        if (string.IsNullOrEmpty(line))
            continue;

        if (!line.StartsWith("//"))
            LinesWithResults.Add(line);
        else if (Engine.Signals.TryGetSignal(line, out var signal))
        {
            diagnoser?.Handle(signal, diagnoserActionParameters);
            process.StandardInput.WriteLine(Engine.Signals.Acknowledgment);
        }
        else
            LinesWithExtraOutput.Add(line);
    }
}
```

There appears to be a couple of Stackoverflow open issues which indicate that they experienced infinite loops in this same manner.

Code Change Proposal:
```csharp
internal void ProcessInput()
{
    string line;
    while (!process.StandardOutput.EndOfStream)
    {
        line = process.StandardOutput.ReadLine();
        if (!string.IsNullOrEmpty(line))
        {
            logger.WriteLine(LogKind.Default, line);

            if (!line.StartsWith("//"))
                LinesWithResults.Add(line);
            else if (Engine.Signals.TryGetSignal(line, out var signal))
            {
                diagnoser?.Handle(signal, diagnoserActionParameters);
                process.StandardInput.WriteLine(Engine.Signals.Acknowledgment);
            }
            else
                LinesWithExtraOutput.Add(line);
        }
    }
}
```
Asynchronous would be considered a better choice and adding a cancellation token on a timer.

#### Original Post
> I started using BenchmarkDotNet yesterday so forgive my ignorance. I clearly have done something wrong. Can't ever get a test to finish. It's always frozen at `// AfterAll`.
> 
> ![image](https://user-images.githubusercontent.com/1414138/42735549-1e9368a4-8824-11e8-83e6-bf7245c34fd2.png)
> 
> It's a NetCore only test, using version 2.1.0.
> 
> I have manipulated parameters, I have manipulated Jobs... It's probably something simple but even my simplest of tests fail to finish.
> 
> I have tried both regular Main and async Main (my preference is async Main).
> 
> ```csharp
> public class Program
> {
>     public static async Task Main(string[] args)
>     {
>         var summary = BenchmarkRunner.Run<Publish>();
> 
>         await Console.Out.WriteLineAsync("CookedRabbit benchmark has finished!");
>     }
> }
> ```
> I have tried [CoreJob], [SimpleJob].
> I have tried simple parameters... just running one iteration example. Nothing seems to work.
> ```csharp
> [MarkdownExporter]
> [SimpleJob(RunStrategy.Monitoring, launchCount: 1, warmupCount: 2, targetCount: 3)]
> public class Publish
> {
>     public Publish()
>     { }
> 
>     RabbitTopologyService _topologyService;
>     RabbitDeliveryService _deliveryService;
>     RabbitMaintenanceService _maintenanceService;
> 
>     public List<byte[]> Payloads { get; set; }
> 
>     public string QueueName { get; set; }
>     public string ExchangeName { get; set; }
> 
>     [GlobalSetup]
>     public void Setup()
>     {
>         QueueName = "CookedRabbit.Benchmark.Scaling";
>         ExchangeName = string.Empty;
> 
>         // Configured for performance.
>         var seasoning = new RabbitSeasoning
>         {
>             ThrottleFastBodyLoops = false,
>             ThrowExceptions = false,
>             WriteErrorsToConsole = false,
>             WriteErrorsToILogger = false,
>             BatchBreakOnException = true
>         };
> 
>         seasoning.PoolSettings.EnableAutoScaling = true;
>         seasoning.PoolSettings.EmptyPoolWaitTime = 10;
>         seasoning.PoolSettings.WriteSleepNoticeToConsole = false;
>         seasoning.PoolSettings.ConnectionPoolCount = 4;
>         seasoning.PoolSettings.ChannelPoolCount = 16;
> 
>         var channelPool = new RabbitChannelPool();
>         channelPool.Initialize(seasoning).GetAwaiter().GetResult();
> 
>         _deliveryService = new RabbitDeliveryService(seasoning, channelPool);
>         _topologyService = new RabbitTopologyService(seasoning, channelPool);
>         _maintenanceService = new RabbitMaintenanceService(seasoning, channelPool);
> 
>         _topologyService.QueueDeclareAsync(QueueName).GetAwaiter().GetResult();
>     }
> 
>     [GlobalCleanup]
>     public void GlobalCleanup()
>     {
>         _maintenanceService.PurgeQueueAsync(QueueName).GetAwaiter().GetResult();
>     }
> 
>     [Params(1, 10, 100, 1000)]
>     public int MessagesToSend { get; set; }
>     [Params(100, 200, 500, 1000)]
>     public int MessageSize;
> 
>     [IterationSetup]
>     public void IterationSetup()
>     {
>         _maintenanceService.PurgeQueueAsync(QueueName).GetAwaiter().GetResult();
>         Payloads = CreatePayloadsAsync(MessagesToSend, MessageSize).GetAwaiter().GetResult();
>     }
> 
>     [IterationCleanup]
>     public void IterationCleanup()
>     {
>         Payloads = new List<byte[]>();
>     }
> 
>     [Benchmark]
>     public async Task Benchmark_Delivery_PublishManyAsync()
>     {
>         await _deliveryService.PublishManyAsync(ExchangeName, QueueName, Payloads, false, null);
>     }
> }
> ```
> 
> Also when stopping the tests, or closing the test, I never see any cleanup done so multiple instances of `dotnet` host are left running.
> ![image](https://user-images.githubusercontent.com/1414138/42735609-11680d28-8825-11e8-9dc0-e2d4a59b6d54.png)
> 
> I was thinking hypothetically, my complicated code may have an issue... Maybe some infinite loop so... I created specifically complex functional and integration tests. All passing with flying colors.
> 
> Any help is much appreciated.  The GitHub repo is here:
> https://github.com/thyams/CookedRabbit

Starting to see some correlation, the dotnet host never finishes / terminates. CodeGen is throwing errors for files being kept open...

Exact same test now fails to build... very curious.

Very peculiar:
```
// Build Exception: Microsoft (R) Build Engine version 15.7.179.6572 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  CookedRabbit.Core.Library -> C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Library\bin\Release\netcoreapp2.1\CookedRabbit.Core.Library.dll
CSC : error CS2012: Cannot open 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' for writing -- 'The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' because it is being used by another process.' [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]

Build FAILED.

CSC : error CS2012: Cannot open 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' for writing -- 'The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' because it is being used by another process.' [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.93


// Build Exception: Microsoft (R) Build Engine version 15.7.179.6572 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  CookedRabbit.Core.Library -> C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Library\bin\Release\netcoreapp2.1\CookedRabbit.Core.Library.dll
CSC : error CS2012: Cannot open 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' for writing -- 'The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' because it is being used by another process.' [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]

Build FAILED.

CSC : error CS2012: Cannot open 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' for writing -- 'The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' because it is being used by another process.' [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:02.02


// Build Exception: Microsoft (R) Build Engine version 15.7.179.6572 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Unable to use package assets cache due to I/O error. This can occur when the same project is built more than once in parallel. Performance may be degraded, but the build result will not be impacted.
  CookedRabbit.Core.Library -> C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Library\bin\Release\netcoreapp2.1\CookedRabbit.Core.Library.dll
c:\program files\dotnet\sdk\2.1.300\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3101: Could not write state file "obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csprojAssemblyReference.cache". The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csprojAssemblyReference.cache' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
CSC : error CS2012: Cannot open 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' for writing -- 'The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' because it is being used by another process.' [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Microsoft.Common.CurrentVersion.targets(5084,5): error MSB3491: Could not write lines to file "obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csproj.FileListAbsolute.txt". The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csproj.FileListAbsolute.txt' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]

Build FAILED.

c:\program files\dotnet\sdk\2.1.300\Microsoft.Common.CurrentVersion.targets(2106,5): warning MSB3101: Could not write state file "obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csprojAssemblyReference.cache". The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csprojAssemblyReference.cache' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
CSC : error CS2012: Cannot open 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' for writing -- 'The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.dll' because it is being used by another process.' [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Microsoft.Common.CurrentVersion.targets(5084,5): error MSB3491: Could not write lines to file "obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csproj.FileListAbsolute.txt". The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\obj\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.csproj.FileListAbsolute.txt' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
    1 Warning(s)
    2 Error(s)

Time Elapsed 00:00:02.05
```

When in doubt... run simple tests again.

Sleep benchmark fails.

```csharp
public class Sleep
{
    [Benchmark]
    public void Time50() => Thread.Sleep(50);

    [Benchmark(Baseline = true)]
    public void Time100() => Thread.Sleep(100);

    [Benchmark]
    public void Time150() => Thread.Sleep(150);
}
```

```
// Build Exception: Microsoft (R) Build Engine version 15.7.179.6572 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  CookedRabbit.Core.Library -> C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Library\bin\Release\netcoreapp2.1\CookedRabbit.Core.Library.dll
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: The "GenerateRuntimeConfigurationFiles" task failed unexpectedly. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: System.IO.IOException: The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\bin\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.runtimeconfig.json' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.File.Create(String path) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteToJsonFile(String fileName, Object value) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteRuntimeConfig(ProjectContext projectContext) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.ExecuteCore() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]

Build FAILED.

c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: The "GenerateRuntimeConfigurationFiles" task failed unexpectedly. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: System.IO.IOException: The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\bin\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.runtimeconfig.json' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.File.Create(String path) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteToJsonFile(String fileName, Object value) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteRuntimeConfig(ProjectContext projectContext) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.ExecuteCore() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.84


// Build Exception: Microsoft (R) Build Engine version 15.7.179.6572 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  CookedRabbit.Core.Library -> C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Library\bin\Release\netcoreapp2.1\CookedRabbit.Core.Library.dll
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: The "GenerateRuntimeConfigurationFiles" task failed unexpectedly. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: System.IO.IOException: The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\bin\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.runtimeconfig.json' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.File.Create(String path) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteToJsonFile(String fileName, Object value) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteRuntimeConfig(ProjectContext projectContext) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.ExecuteCore() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]

Build FAILED.

c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: The "GenerateRuntimeConfigurationFiles" task failed unexpectedly. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018: System.IO.IOException: The process cannot access the file 'C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\bin\Release\netcoreapp2.1\CookedRabbit.Core.Benchmark.runtimeconfig.json' because it is being used by another process. [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at System.IO.File.Create(String path) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteToJsonFile(String fileName, Object value) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.WriteRuntimeConfig(ProjectContext projectContext) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateRuntimeConfigurationFiles.ExecuteCore() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
c:\program files\dotnet\sdk\2.1.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(171,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\GitHub\personal\CookedRabbit\NetCore\CookedRabbit.Core.Benchmark\CookedRabbit.Core.Benchmark.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.90
```

--- a/src/BenchmarkDotNet/Helpers/ConsoleExitHandler.cs
+++ b/src/BenchmarkDotNet/Helpers/ConsoleExitHandler.cs
@@ -57,7 +57,7 @@ namespace BenchmarkDotNet.Helpers
         // the user has closed the console window so we kill the entire process tree
         private void ProcessExitEventHandlerHandlerCallback(object sender, EventArgs e) => KillProcessTree();
 
-        private void KillProcessTree()
+        internal void KillProcessTree()
         {
             try
             {

--- a/src/BenchmarkDotNet/Loggers/SynchronousProcessOutputLoggerWithDiagnoser.cs
+++ b/src/BenchmarkDotNet/Loggers/SynchronousProcessOutputLoggerWithDiagnoser.cs
@@ -55,6 +55,13 @@ namespace BenchmarkDotNet.Loggers
                 {
                     diagnoser?.Handle(signal, diagnoserActionParameters);
                     process.StandardInput.WriteLine(Engine.Signals.Acknowledgment);
+
+                    if (signal == HostSignal.AfterAll)
+                    {
+                        // we have received the last signal so we can stop reading the output
+                        // if the process won't exit after this, its hung and needs to be killed
+                        return;
+                    }
                 }
                 else if (!string.IsNullOrEmpty(line))
                 {

--- a/src/BenchmarkDotNet/Parameters/ParameterInstance.cs
+++ b/src/BenchmarkDotNet/Parameters/ParameterInstance.cs
@@ -8,7 +8,7 @@ using JetBrains.Annotations;
 
 namespace BenchmarkDotNet.Parameters
 {
-    public class ParameterInstance
+    public class ParameterInstance : IDisposable
     {
         public const string NullParameterTextRepresentation = "?";
 
@@ -23,6 +23,8 @@ namespace BenchmarkDotNet.Parameters
             this.value = value;
             maxParameterColumnWidth = summaryStyle?.MaxParameterColumnWidth ?? SummaryStyle.DefaultMaxParameterColumnWidth;
         }
+
+        public void Dispose() => (Value as IDisposable)?.Dispose();
 
         public string Name => Definition.Name;
         public bool IsStatic => Definition.IsStatic;

--- a/src/BenchmarkDotNet/Parameters/ParameterInstances.cs
+++ b/src/BenchmarkDotNet/Parameters/ParameterInstances.cs
@@ -1,10 +1,11 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
 using System.Linq;
 using BenchmarkDotNet.Extensions;
 
 namespace BenchmarkDotNet.Parameters
 {
-    public class ParameterInstances
+    public class ParameterInstances : IDisposable
     {
         public IReadOnlyList<ParameterInstance> Items { get; }
         public int Count => Items.Count;
@@ -16,6 +17,14 @@ namespace BenchmarkDotNet.Parameters
         public ParameterInstances(IReadOnlyList<ParameterInstance> items)
         {
             Items = items;
+        }
+
+        public void Dispose()
+        {
+            foreach (var parameterInstance in Items)
+            {
+                parameterInstance.Dispose();
+            }
         }
 
         public string FolderInfo => string.Join("_", Items.Select(p => $"{p.Name}-{p.ToDisplayText()}")).AsValidFileName();

--- a/src/BenchmarkDotNet/Parameters/SmartParamBuilder.cs
+++ b/src/BenchmarkDotNet/Parameters/SmartParamBuilder.cs
@@ -1,4 +1,6 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
 using System.Reflection;
@@ -97,9 +99,8 @@ namespace BenchmarkDotNet.Parameters
                 ? $"[{argumentIndex}]" // IEnumerable<object[]>
                 : string.Empty; // IEnumerable<object>
 
-            // we just execute (cast)source.ToArray()[case][argumentIndex];
-            // we know that source is IEnumerable so we can do that!
-            return $"{cast}System.Linq.Enumerable.ToArray({source.Name}{callPostfix})[{sourceIndex}]{indexPostfix};";
+            // we do something like enumerable.ElementAt(sourceIndex)[argumentIndex];
+            return $"{cast}BenchmarkDotNet.Parameters.ParameterExtractor.GetParameter({source.Name}{callPostfix}, {sourceIndex}){indexPostfix};";
         }
     }
 
@@ -129,8 +130,38 @@ namespace BenchmarkDotNet.Parameters
 
             string callPostfix = source is PropertyInfo ? string.Empty : "()";
 
-            // we just execute (cast)source.ToArray()[index];
-            return $"{cast}System.Linq.Enumerable.ToArray({instancePrefix}.{source.Name}{callPostfix})[{index}];";
+            // we so something like enumerable.ElementAt(index);
+            return $"{cast}BenchmarkDotNet.Parameters.ParameterExtractor.GetParameter({instancePrefix}.{source.Name}{callPostfix}, {index});";
+        }
+    }
+
+    public static class ParameterExtractor
+    {
+        [EditorBrowsable(EditorBrowsableState.Never)] // hide from intellisense, it's public so we can call it form the boilerplate code
+        public static T GetParameter<T>(IEnumerable<T> parameters, int index)
+        {
+            int count = 0;
+
+            foreach (T parameter in parameters)
+            {
+                if (count == index)
+                {
+                    return parameter;
+                }
+
+                if (parameter is IDisposable disposable)
+                {
+                    // parameters might contain locking finalizers which might cause the benchmarking process to hung at the end
+                    // to avoid that, we dispose the parameters that were created, but won't be used
+                    // (for every test case we have to enumerate the underlying source enumerator and stop when we reach index of given test case)
+                    // See https://github.com/dotnet/BenchmarkDotNet/issues/1383 and https://github.com/dotnet/runtime/issues/314 for more
+                    disposable.Dispose();
+                }
+
+                count++;
+            }
+
+            throw new InvalidOperationException("We should never get here!");
         }
     }
 }

--- a/src/BenchmarkDotNet/Running/BenchmarkCase.cs
+++ b/src/BenchmarkDotNet/Running/BenchmarkCase.cs
@@ -6,7 +6,7 @@ using BenchmarkDotNet.Parameters;
 
 namespace BenchmarkDotNet.Running
 {
-    public class BenchmarkCase : IComparable<BenchmarkCase>
+    public class BenchmarkCase : IComparable<BenchmarkCase>, IDisposable
     {
         public Descriptor Descriptor { get; }
         public Job Job { get; }
@@ -25,6 +25,8 @@ namespace BenchmarkDotNet.Running
             Parameters = parameters;
             Config = config;
         }
+
+        public void Dispose() => Parameters.Dispose();
 
         public int CompareTo(BenchmarkCase other) => string.Compare(FolderInfo, other.FolderInfo, StringComparison.Ordinal);
 

--- a/src/BenchmarkDotNet/Running/BenchmarkRunInfo.cs
+++ b/src/BenchmarkDotNet/Running/BenchmarkRunInfo.cs
@@ -3,13 +3,21 @@ using BenchmarkDotNet.Configs;
 
 namespace BenchmarkDotNet.Running
 {
-    public class BenchmarkRunInfo
+    public class BenchmarkRunInfo : IDisposable
     {
         public BenchmarkRunInfo(BenchmarkCase[] benchmarksCase, Type type, ImmutableConfig config)
         {
             BenchmarksCases = benchmarksCase;
             Type = type;
             Config = config;
+        }
+
+        public void Dispose()
+        {
+            foreach (var benchmarkCase in BenchmarksCases)
+            {
+                benchmarkCase.Dispose();
+            }
         }
 
         public BenchmarkCase[] BenchmarksCases { get; }

--- a/src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
+++ b/src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
@@ -501,10 +501,12 @@ namespace BenchmarkDotNet.Running
                 logger.WriteLineError($"Executable {buildResult.ArtifactsPaths.ExecutablePath} not found");
             }
 
-            if (executeResult.ExitCode != 0)
+            // exit code can be different than 0 if the process has hanged at the end
+            // so we check if some results were reported, if not then it was a failure
+            if (executeResult.ExitCode != 0 && executeResult.Data.IsEmpty())
             {
                 success = false;
-                logger.WriteLineError("ExitCode != 0");
+                logger.WriteLineError("ExitCode != 0 and no results reported");
             }
 
             return executeResult;

--- a/src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
+++ b/src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
@@ -107,6 +107,14 @@ namespace BenchmarkDotNet.Running
                 }
                 finally
                 {
+                    // some benchmarks might be using parameters that have locking finalizers
+                    // so we need to dispose them after we are done running the benchmarks
+                    // see https://github.com/dotnet/BenchmarkDotNet/issues/1383 and https://github.com/dotnet/runtime/issues/314 for more
+                    foreach (var benchmarkInfo in benchmarkRunInfos)
+                    {
+                        benchmarkInfo.Dispose();
+                    }
+
                     compositeLogger.WriteLineHeader("// * Artifacts cleanup *");
                     Cleanup(new HashSet<string>(artifactsToCleanup.Distinct()));
                     compositeLogger.Flush();

--- a/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliExecutor.cs
+++ b/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliExecutor.cs
@@ -69,7 +69,7 @@ namespace BenchmarkDotNet.Toolchains.DotNetCli
             startInfo.SetEnvironmentVariables(benchmarkCase, resolver);
 
             using (var process = new Process { StartInfo = startInfo })
-            using (new ConsoleExitHandler(process, logger))
+            using (var consoleExitHandler = new ConsoleExitHandler(process, logger))
             {
                 var loggerWithDiagnoser = new SynchronousProcessOutputLoggerWithDiagnoser(logger, process, diagnoser, benchmarkCase, benchmarkId);
 
@@ -87,14 +87,14 @@ namespace BenchmarkDotNet.Toolchains.DotNetCli
 
                 loggerWithDiagnoser.ProcessInput();
 
-                process.WaitForExit(); // should we add timeout here?
-
-                if (process.ExitCode == 0)
+                if (!process.WaitForExit(milliseconds: 250))
                 {
-                    return new ExecuteResult(true, process.ExitCode, process.Id, loggerWithDiagnoser.LinesWithResults, loggerWithDiagnoser.LinesWithExtraOutput);
+                    logger.WriteLineInfo("// The benchmarking process did not quit on time, it's going to get force killed now.");
+
+                    consoleExitHandler.KillProcessTree();
                 }
 
-                return new ExecuteResult(true, process.ExitCode, process.Id, Array.Empty<string>(), Array.Empty<string>());
+                return new ExecuteResult(true, process.ExitCode, process.Id, loggerWithDiagnoser.LinesWithResults, loggerWithDiagnoser.LinesWithExtraOutput);
             }
         }
     }

--- a/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliExecutor.cs
+++ b/src/BenchmarkDotNet/Toolchains/DotNetCli/DotNetCliExecutor.cs
@@ -86,18 +86,12 @@ namespace BenchmarkDotNet.Toolchains.DotNetCli
                 }
 
                 loggerWithDiagnoser.ProcessInput();
-                string standardError = process.StandardError.ReadToEnd();
 
                 process.WaitForExit(); // should we add timeout here?
 
                 if (process.ExitCode == 0)
                 {
                     return new ExecuteResult(true, process.ExitCode, process.Id, loggerWithDiagnoser.LinesWithResults, loggerWithDiagnoser.LinesWithExtraOutput);
-                }
-
-                if (!string.IsNullOrEmpty(standardError))
-                {
-                    logger.WriteError(standardError);
                 }
 
                 return new ExecuteResult(true, process.ExitCode, process.Id, Array.Empty<string>(), Array.Empty<string>());

--- a/src/BenchmarkDotNet/Toolchains/Executor.cs
+++ b/src/BenchmarkDotNet/Toolchains/Executor.cs
@@ -40,13 +40,13 @@ namespace BenchmarkDotNet.Toolchains
             try
             {
                 using (var process = new Process { StartInfo = CreateStartInfo(benchmarkCase, artifactsPaths, args, resolver) })
-                using (new ConsoleExitHandler(process, logger))
+                using (var consoleExitHandler = new ConsoleExitHandler(process, logger))
                 {
                     var loggerWithDiagnoser = new SynchronousProcessOutputLoggerWithDiagnoser(logger, process, diagnoser, benchmarkCase, benchmarkId);
 
                     diagnoser?.Handle(HostSignal.BeforeProcessStart, new DiagnoserActionParameters(process, benchmarkCase, benchmarkId));
 
-                    return Execute(process, benchmarkCase, loggerWithDiagnoser, logger);
+                    return Execute(process, benchmarkCase, loggerWithDiagnoser, logger, consoleExitHandler);
                 }
             }
             finally
@@ -55,7 +55,7 @@ namespace BenchmarkDotNet.Toolchains
             }
         }
 
-        private ExecuteResult Execute(Process process, BenchmarkCase benchmarkCase, SynchronousProcessOutputLoggerWithDiagnoser loggerWithDiagnoser, ILogger logger)
+        private ExecuteResult Execute(Process process, BenchmarkCase benchmarkCase, SynchronousProcessOutputLoggerWithDiagnoser loggerWithDiagnoser, ILogger logger, ConsoleExitHandler consoleExitHandler)
         {
             logger.WriteLineInfo($"// Execute: {process.StartInfo.FileName} {process.StartInfo.Arguments} in {process.StartInfo.WorkingDirectory}");
 
@@ -69,17 +69,17 @@ namespace BenchmarkDotNet.Toolchains
 
             loggerWithDiagnoser.ProcessInput();
 
-            process.WaitForExit(); // should we add timeout here?
-
-            if (process.ExitCode == 0)
+            if (!process.WaitForExit(milliseconds: 250))
             {
-                return new ExecuteResult(true, process.ExitCode, process.Id, loggerWithDiagnoser.LinesWithResults, loggerWithDiagnoser.LinesWithExtraOutput);
+                logger.WriteLineInfo("// The benchmarking process did not quit on time, it's going to get force killed now.");
+
+                consoleExitHandler.KillProcessTree();
             }
 
             if (loggerWithDiagnoser.LinesWithResults.Any(line => line.Contains("BadImageFormatException")))
                 logger.WriteLineError("You are probably missing <PlatformTarget>AnyCPU</PlatformTarget> in your .csproj file.");
 
-            return new ExecuteResult(true, process.ExitCode, process.Id, Array.Empty<string>(), Array.Empty<string>());
+            return new ExecuteResult(true, process.ExitCode, process.Id, loggerWithDiagnoser.LinesWithResults, loggerWithDiagnoser.LinesWithExtraOutput);
         }
 
         private ProcessStartInfo CreateStartInfo(BenchmarkCase benchmarkCase, ArtifactsPaths artifactsPaths, string args, IResolver resolver)

--- a/tests/BenchmarkDotNet.IntegrationTests/ArgumentsTests.cs
+++ b/tests/BenchmarkDotNet.IntegrationTests/ArgumentsTests.cs
@@ -770,6 +770,56 @@ namespace BenchmarkDotNet.IntegrationTests
             }
         }
 
-        private void CanExecute<T>(IToolchain toolchain) => CanExecute<T>(CreateSimpleConfig(job: Job.Dry.With(toolchain)));
+        [Fact]
+        public void UnusedDisposableParamsAreDisposed() => CanExecute<WithDisposableArguments>(Job.Default.GetToolchain());
+
+        public class WithDisposableArguments
+        {
+            public IEnumerable<Disposable> GetDisposables()
+            {
+                yield return new Disposable(0);
+                yield return new Disposable(1);
+            }
+
+            [ParamsSource(nameof(GetDisposables))]
+            public Disposable used;
+
+            [Benchmark]
+            public void CheckDisposed()
+            {
+                if (used.Id == 0)
+                {
+                    if (Disposable.Created != 1)
+                        throw new ArgumentException("Only one instance should be created so far!");
+                    if (Disposable.Disposed != 0)
+                        throw new ArgumentException("None should be disposed as only one was created and is still in use");
+                }
+                if (used.Id == 1)
+                {
+                    if (Disposable.Created != 2)
+                        throw new ArgumentException("Two instances should be created so far!");
+                    if (Disposable.Disposed != 1)
+                        throw new ArgumentException("The first one should be disposed as it's not used");
+                }
+            }
+
+            public class Disposable : IDisposable
+            {
+                public static int Created = 0;
+                public static int Disposed = 0;
+
+                public int Id { get; private set; }
+
+                public Disposable(int id)
+                {
+                    Id = id;
+                    ++Created;
+                }
+
+                public void Dispose() => ++Disposed;
+            }
+        }
+
+        private void CanExecute<T>(IToolchain toolchain) => CanExecute<T>(CreateSimpleConfig(job: Job.Dry.WithToolchain(toolchain)));
     }
 }

